pilot/pkg/controllers/untaint/nodeuntainter.go

```diff
@@ -0,0 +1,240 @@
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/krt"
+	istiolog "istio.io/istio/pkg/log"
+)
+
+var log = istiolog.RegisterScope("untaint", "CNI node-untaint controller")
+
+const (
+	TaintName = "cni.istio.io/not-ready"
+)
+
+var istioCniLabels = map[string]string{
+	"k8s-app": "istio-cni-node",
+}
+
+type NodeUntainter struct {
+	podsClient  kclient.Client[*v1.Pod]
+	nodesClient kclient.Client[*v1.Node]
+	cnilabels   labels.Instance
+	ourNs       string
+	queue       controllers.Queue
+}
+
+func filterNamespace(ns string) func(any) bool {
+	return func(obj any) bool {
+		object := controllers.ExtractObject(obj)
+		if object == nil {
+			return false
+		}
+		return ns == object.GetNamespace()
+	}
+}
+
+func NewNodeUntainter(stop <-chan struct{}, kubeClient kubelib.Client, cniNs, sysNs string) *NodeUntainter {
+	log.Debugf("starting node untainter with labels %v", istioCniLabels)
+	ns := cniNs
+	if ns == "" {
+		ns = sysNs
+	}
+	podsClient := kclient.NewFiltered[*v1.Pod](kubeClient, kclient.Filter{
+		ObjectFilter:    filterNamespace(ns),
+		ObjectTransform: kubelib.StripPodUnusedFields,
+	})
+	nodes := kclient.NewFiltered[*v1.Node](kubeClient, kclient.Filter{ObjectTransform: kubelib.StripNodeUnusedFields})
+	nt := &NodeUntainter{
+		podsClient:  podsClient,
+		nodesClient: nodes,
+		cnilabels:   labels.Instance(istioCniLabels),
+		ourNs:       ns,
+	}
+	nt.setup(stop)
+	return nt
+}
+
+func (n *NodeUntainter) setup(stop <-chan struct{}) {
+	nodes := krt.WrapClient[*v1.Node](n.nodesClient)
+	pods := krt.WrapClient[*v1.Pod](n.podsClient)
+
+	readyCniPods := krt.NewCollection(pods, func(ctx krt.HandlerContext, p *v1.Pod) *v1.Pod {
+		log.Debugf("cniPods event: %s", p.Name)
+		if p.Namespace != n.ourNs {
+			return nil
+		}
+		if !n.cnilabels.SubsetOf(p.ObjectMeta.Labels) {
+			return nil
+		}
+		if !IsPodReadyConditionTrue(p.Status) {
+			return nil
+		}
+		log.Debugf("pod %s on node %s ready!", p.Name, p.Spec.NodeName)
+		return p
+	}, krt.WithStop(stop))
+
+	// these are all the nodes that have a ready cni pod. if the cni pod is ready,
+	// it means we are ok scheduling pods to it.
+	readyCniNodes := krt.NewCollection(readyCniPods, func(ctx krt.HandlerContext, p v1.Pod) *v1.Node {
+		pnode := krt.FetchOne(ctx, nodes, krt.FilterKey(p.Spec.NodeName))
+		if pnode == nil {
+			return nil
+		}
+		node := *pnode
+		if !hasTaint(node) {
+			return nil
+		}
+		return node
+	}, krt.WithStop(stop))
+
+	n.queue = controllers.NewQueue("untaint nodes",
+		controllers.WithReconciler(n.reconcileNode),
+		controllers.WithMaxAttempts(5))
+
+	// remove the taints from readyCniNodes
+	readyCniNodes.Register(func(o krt.Event[v1.Node]) {
+		if o.Event == controllers.EventDelete {
+			return
+		}
+		if o.New != nil {
+			log.Debugf("adding node to queue event: %s", o.New.Name)
+			n.queue.AddObject(o.New)
+		}
+	})
+}
+
+func (n *NodeUntainter) Run(stop <-chan struct{}) {
+	n.queue.Run(stop)
```

Comment:
`{podsClient/nodesClient}.ShutdownHandlers()` would be good.

its more important for things with leader election, since they can start and stop